dev/fattest.simplicity/src/componenttest/security/utils/SSLUtils.java

```diff
@@ -0,0 +1,175 @@
+        generator.initialize(2048, new SecureRandom());
+        KeyPair pair = generator.generateKeyPair();
+
+        return pair;
+    }
+
+    public static Certificate selfSign(KeyPair keyPair, String subjectDN, List<String> genericNameList)
+            throws Exception
+    {
+        Log.info(c, "selfSign", "entering SSLUtils.selfSign");
+
+        final String DNS_PATTERN = "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9])$";
+        final String IPV4_PATTERN = "^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\\.(?!$)|$)){4}$";
+
+        Provider bcProvider = new BouncyCastleProvider();
+        Security.addProvider(bcProvider);
+
+        long now = System.currentTimeMillis();
+        Date startDate = new Date(now);
+
+        X500Name dnName = new X500Name(subjectDN);
+
+        // Using the current timestamp as the certificate serial number
+        BigInteger certSerialNumber = new BigInteger(Long.toString(now));
+
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(startDate);
+        // 1 Yr validity
+        calendar.add(Calendar.YEAR, 1);
+
+        Date endDate = calendar.getTime();
+
+        // Use appropriate signature algorithm based on your keyPair algorithm.
+        String signatureAlgorithm = "SHA256WithRSA";
+
+        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(keyPair
+                .getPublic().getEncoded());
+
+        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(dnName,
+                certSerialNumber, startDate, endDate, dnName, subjectPublicKeyInfo);
+
+        List<GeneralName> altNames = new ArrayList<GeneralName>();
+        for (String name : genericNameList) {
+            if (name.matches(DNS_PATTERN)) {
+                Log.info(c, "selfSign", "adding dns GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            } else if (name.matches(IPV4_PATTERN)) {
+                Log.info(c, "selfSign", "adding ip GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            }
+        }
+
+        GeneralNames subjectAltNames = GeneralNames.getInstance(new DERSequence((GeneralName[]) altNames.toArray(new GeneralName[] {})));
+        certificateBuilder.addExtension(Extension.subjectAlternativeName, false, subjectAltNames);
+
+        ContentSigner contentSigner = new JcaContentSignerBuilder(signatureAlgorithm).setProvider(
+                bcProvider).build(keyPair.getPrivate());
+
+        X509CertificateHolder certificateHolder = certificateBuilder.build(contentSigner);
+
+        Certificate selfSignedCert = new JcaX509CertificateConverter()
+                .getCertificate(certificateHolder);
+
+        return selfSignedCert;
+    }
+
+    public static void exportPublicKeyToFile(String path, KeyPair keyPair) throws Exception {
```

Comment:
In general, when writing utility methods for working with files, we should pass in the `File` or `Path` object, rather than passing in a string.

Any requirements on the file passed in (like it must not exist, or its parent directory must exist) should be documented. If you're doing anything else in your method (like automatically creating parent directories), that must be documented.